packages/arb-avm-cpp/data_storage/src/arbcore.cpp

```diff
@@ -548,56 +556,35 @@ rocksdb::Status ArbCore::reorgCheckpoints(
-                    // in database
-                    if (coreConfig.debug) {
-                        std::cerr << "Last L2 block saved to database: "
-                                  << checkpoint.output.l2_block_number
+                if (finished) {
+                    if (initial_start && coreConfig.debug) {
+                        std::cerr << "Loading L2 block saved to "
+                                     "database: "
+                                  << machine_output.l2_block_number
                                   << std::endl;
                     }
-                    selected_machine_output = checkpoint.output;
-                }
-                if (checkpoint.output.arb_gas_used == 0 ||
-                    check_output(checkpoint.output)) {
-                    // All outdated checkpoints have been removed
-                    if (isValid(tx, checkpoint.output.fully_processed_inbox)) {
-                        // Good checkpoint
-                        if (initial_start && coreConfig.debug) {
-                            std::cerr << "Loading L2 block saved to "
-                                         "database: "
-                                      << checkpoint.output.l2_block_number
-                                      << std::endl;
-                        }
-                        auto mach = combined_sideload_cache.atOrBeforeGas(
-                            checkpoint.output.arb_gas_used, std::nullopt,
-                            std::nullopt, coreConfig.checkpoint_load_gas_cost,
-                            std::numeric_limits<uint256_t>::max());
-                        if (mach.machine != nullptr) {
-                            // Found machine in cache
-                            setup = std::make_unique<MachineThread>(
-                                mach.machine->machine_state);
-                            break;
-                        }
 
-                        try {
-                            // Load machine from database
-                            setup = getMachineUsingStateKeys<MachineThread>(
-                                tx, checkpoint, cache,
-                                coreConfig.lazy_load_core_machine);
-                            break;
-                        } catch (const std::exception& e) {
-                            std::cerr
-                                << "Error loading machine from checkpoint: "
-                                << e.what() << std::endl;
-                            assert(false);
-                        }
-                    } else {
-                        std::cerr
-                            << "Unexpectedly invalid checkpoint inbox at "
-                               "message count "
-                            << checkpoint.output.fully_processed_inbox.count
-                            << std::endl;
+                    auto mach = combined_machine_cache.atOrBeforeGas(
```

Comment:
I think this should contain the database gas used as an option, though the database load gas cost should be much higher than in non-core thread loading if `lazy_load_core_machine` is false since the cost is much higher when greedy loading. This should also be the only call to `atOrBeforeGas` I think

Reply:
I also think we might need to use the `isValid` check on machines returned from the cache since that's how we ensure a machine isn't invalidated by an inbox reorg. That said I'm not totally sure if that's necessary post removal of the inbox peek field @PlasmaPower?

Reply:
Without inbox peek the cache reorg should be reliable. That said, I also don't think `isValid` is very expensive, so we could check it just in case.